scheduler/src/ratelimit/token_bucket_filter.clj

```diff
@@ -0,0 +1,88 @@
+;; limitations under the License.
+;;
+(ns ratelimit.token-bucket-filter)
+
+(defn create-tbf
+  "Make a token bucket filter (tbf), with a given refill rate (units per timestep), a maximum token capacity, and give it the
+  current timestamp.
+
+  This token bucket filter is a bit unusual, instead of earning tokens that you can later spend, we allow the current
+  tokens to be negative. Thus, you can (in one request) spend an unbounded number of tokens. However, you will be
+  in debt until your token balance is at least 0. This design allows the tokens in a single request to exceed the max-tokens
+  in the tbf. (If we didn't do it this way, max-tokens is an implicit maximum limit on request size.) This way
+  we can independently pick the burst size in the tbf and the maximum request size.
+
+  This library does not handle blocking or other operations. Typical workflows are to:
+
+  ** Admission control **
+  This is good when you don't know how big a request will be until afterwards.
+
+  Use in-debt? as part of request admission control, rejecting a request if it returns true. When a request completes (or is aborted),
+  take the tokens that it used, affecting the next request.
+
+  ** Delaying requests **
+  If you want to delay requests instead of rejecting. Take the tokens the request needs, pause for the returned pause time.
+  (E.g., schedule the message to be sent at that particular pause time.
+  "
+  ([^double token-rate ^long max-tokens ^long current-time ^long current-tokens]
+   {:pre [(> token-rate 0)
+          (>= max-tokens 0)
+          (>= current-time 0)]}
+   {:token-rate token-rate :max-tokens max-tokens :last-update current-time :current-tokens current-tokens})
+  ([^double token-rate ^long max-tokens ^long current-time]
+   (create-tbf token-rate max-tokens current-time 0)))
+
+(defn- add-tokens-helper
+  "Add tokens (if there are any). If less than a whole token added, keep the current state. Returns the current state.
+  Enforces max-tokens limit. Allows tokens to be removed or added. If no timestamp supplied, just alters the token count."
+  ([{:keys [current-tokens max-tokens] :as tbf} ^long tokens-to-dispense ^long current-time]
+   (if (zero? tokens-to-dispense)
+     ; Only dispense when we have at least a whole token to add on.
+     tbf
+     (-> tbf
+         (assoc :last-update current-time)
+         (update :current-tokens #(-> % (+ tokens-to-dispense) (min max-tokens))))))
+  ([tbf ^long tokens-to-dispense]
+   (add-tokens-helper tbf tokens-to-dispense (:last-update tbf))))
+
+(defn earn-tokens
+  "Earn tokens. Helper function, takes the current timestamp and adds in any newly earned tokens. It is numerically
+  robust to the timestamp being out of order."
+  [{:keys [last-update token-rate] :as tbf} ^long current-time]
+  {:pre [(>= current-time 0)]}
+  (let [current-time (max current-time last-update)]
+    (cond-> tbf
+            (not (= current-time last-update))
```

Comment:
Indentation appears off.  `cond->`  indents to two whitespaces on the new line.

Reply:
cursive indents it that way, maybe aligning the lines?

Reply:
You can configure individual forms to use different indent levels. https://cursive-ide.com/userguide/formatting.html

Reply:
Yes. Other code in the file has hte same indent.